twitter_clone/main.py

```diff
@@ -25,5 +54,231 @@ def decorated_function(*args, **kwargs):
+            g.db.execute(query, (u_id, tweet))
+            g.db.commit()
+            flash('Tweet posted successfully')
+            flash_type = SUCCESS
+            
+    # Get all tweet data for this user
+    query = '''SELECT u.username, t.id, t.created, t.content, u.id 
+        FROM user as u INNER JOIN tweet as t
+        ON u.id=t.user_id
+        WHERE u.username=?
+        ORDER BY t.created DESC;'''
+    cursor = g.db.execute(query, (username,))
+    tweets = cursor.fetchall()
+    
+    # Display the user's twitter feed page
+    return render_template('feed.html', own=own, tweets=tweets, flash_type=flash_type)
+
+@app.route('/tweets/<t_id>/delete', methods=['POST'])
+# @login_required
+def delete(t_id):
+    
+    # Get information about the tweet to be deleted
+    query = '''SELECT t.user_id, u.username
+        FROM tweet AS t INNER JOIN user AS u 
+        ON t.user_id=u.id
+        WHERE t.id=?;'''
+    
+    # If the tweet id given is not valid or no such tweet exists, return a 404
+    try:
+        cursor = g.db.execute(query, (t_id,))
+        results = cursor.fetchall()
+        # from pprint import pprint
+        # print('Invalid results:')
+        # pprint(results)
+        assert results
+    except:
+        abort(404) # TODO: Customize this?
+    
+    # Verify that the user deleting a tweet is logged in
+    if 'user_id' not in session:
+        return redirect(url_for('login'))
+    
+    # Determine whose tweet is being deleted
+    u_id = results[0][0]
+    username = results[0][1]
+    
+    # Verify that the authenticated user is deleting his own tweet
+    if u_id != session['user_id']:
+        flash('''You can't just go around deleting other people's tweets!
+            How would you feel if someone deleted one of your tweets?''')
+    
+    # Delete the tweet
+    else:
+        query = '''DELETE FROM tweet WHERE id=?;'''
+        g.db.execute(query, (int(t_id),))
+        g.db.commit()
+        
+    # Redirect to the user's feed
+    # return redirect(url_for('feed', username=username))
+    return redirect(url_for('base'))
+
+@app.route('/profile', methods=['GET', 'POST'])
+@login_required
+def profile():
+    '''
+    Obtain the user details and pass to the profile templage for reply.
+    '''
+    error = None 
+    flash_type = SUCCESS
+    # return render_template('login.html')
+    query = '''SELECT username, first_name, last_name, birth_date 
+        FROM user
+        WHERE id=?;'''
+    cursor = g.db.execute(query, (str(session['user_id']),))
+    results = cursor.fetchone()
+    (user, fn, ln, dob) = results
+
+    if request.method == 'POST':
+
+        # If the user supplies an invalid date, error out and don't save
+        if request.form['birth_date'] and not _date_check(request.form['birth_date']):
+            flash('Invalid date')
+            flash_type = DANGER
+            
+        # If the user supplies no date or a valid date, save the updates
```

Comment:
Really great code overall - I learned a lot reading through it! 

Code looks very clean and all the validation of user input is impressive.